cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -233,63 +237,75 @@ public ApplicationMeta getApplication(String namespaceId, String appId, String v
-   *
-   * @param cursor the cursor position to start the scan. A cursor can be obtained
-   *               from the call to the given {@link BiConsumer} for some previous scan, or use
-   *               {@link Cursor#EMPTY} to start a scan at the beginning
-   * @param func a {@link BiFunction} to consume application metadata entries generated by the scan. The boolean
-   *             value returned is {@code true}, the scan will continue; otherwise the scan will stop and return.
-   * @throws IOException if failed to scan the storage
-   */
-  public void scanApplications(Cursor cursor,
-                               BiFunction<Cursor, Map.Entry<ApplicationId, ApplicationMeta>, Boolean> func)
-    throws IOException {
-
-    Range range = Range.create(cursor.fields, cursor.bound,
-                               Collections.emptyList(), Range.Bound.INCLUSIVE);
-
-    scanApplicationsWithRange(range, func);
-  }
-
   /**
    * Scans applications starting with the given cursor position for a given namespace.
    *
-   * @param namespaceId the namespaceId for which the scan is being done.
-   * @param cursor the cursor position to start the scan. A cursor can be obtained
-   *               from the call to the given {@link BiConsumer} for some previous scan, or use
-   *               {@link Cursor#EMPTY} to start a scan at the beginning
+   * @param request parameters defining filters, sorting, limits
+   * @param cursor the cursor position to start the scan.
+   *               Cursor overrides {@link ScanApplicationsRequest#getRestartAfter()} if non-empty.
+   *               A cursor can be obtained from the call to the given {@link BiConsumer} for some previous scan.
+   *               If {@link Cursor#EMPTY} is passed, scan starts from {@link ScanApplicationsRequest#getRestartAfter()}
+   *               or from beginning if it's null.
    * @param func a {@link BiFunction} to consume application metadata entries generated by the scan. The boolean
    *             value returned is {@code true}, the scan will continue; otherwise the scan will stop and return.
    * @throws IOException if failed to scan the storage
    */
-  public void scanApplications(String namespaceId, Cursor cursor,
+  public void scanApplications(ScanApplicationsRequest request, Cursor cursor,
                                BiFunction<Cursor, Map.Entry<ApplicationId, ApplicationMeta>, Boolean> func)
     throws IOException {
 
+    if (cursor.fields.isEmpty() && request.getRestartAfter() != null) {
+      if (request.getNamespaceId() != null &&
+        !request.getNamespaceId().equals(request.getRestartAfter().getNamespaceId())) {
+        throw new IllegalArgumentException("Requested to restart scan from application " + request.getRestartAfter() +
+          " that is outside of scan namespace " + request.getNamespaceId()
+        );
+      }
+      cursor = new Cursor(getApplicationPrimaryKeys(request.getRestartAfter()), Range.Bound.EXCLUSIVE);
```

Comment:
looks like cursor and `restartAfter` are interchangeable, why do we need both ?

Reply:
I would suggest we have two `scanApplications` method,
the 1st one only accept request and the 2nd one accept filters, cursor , order 
and the 1st can call second one by converting restartAfter to cursor. 
Then we dont' need to have those logic that defines the relationship between restartAfter and cursor if both are passed.

Reply:
Mostly usage / layering. I can remove Cursor, but this would make the batching code create an ApplicationId out of cursor each time and then recreate the request with the new cursor. 
Also  restartAfter is mostly for highler layers / UI to do pagination and Cursor is for batching. So while they are filtering on same criteria they serve different purpose. 
Another option is to have a separate Request class on API level (Store) and implementation level (AppMetadataStore), but IMHO this would be an overkill.

Reply:
Actually, done. I removed the cursor for application scans altogether